src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+        p-func-b6
+        p-func-b7
+        p-func-b8
+        p-func-b9
+        p-func-b10
+        p-func-b11
+        p-func-b12
+        p-func-b13
+        p-func-b14
+        p-func-b15
+        p-func-b16))
+      (base p-func-base)
+      (coeff p-func-coeff)
+      (nonalpha-discount p-func-non-alpha-discount)
+      (no-vowel-discount p-func-no-vowel-discount))))
+    (let (
+      (preorder (unwrap!
+        (map-get? namespace-preorders ((hashed-salted-namespace hashed-salted-namespace) (buyer contract-caller)))
+        (err ERR_NAMESPACE_PREORDER_NOT_FOUND)))
+      (namespace-price (unwrap! 
+        (compute-namespace-price? namespace)
+        (err ERR_NAMESPACE_BLANK))))
+    ;; The namespace must only have valid chars
+    (asserts!
+      (not (has-invalid-chars namespace))
+      (err ERR_NAMESPACE_CHARSET_INVALID))
+    ;; The namespace must not exist yet in the `namespaces` table
+    (asserts!
+      (is-none (map-get? namespaces ((namespace namespace))))
+      (err ERR_NAMESPACE_ALREADY_EXISTS))
+    ;; The amount burnt must be equal to or greater than the cost of the namespace
+    (asserts!
+      (>= (get stx-burned preorder) namespace-price)
+      (err ERR_NAMESPACE_STX_BURNT_INSUFFICIENT))
+    ;; This transaction must arrive within 24 hours of its `NAMESPACE_PREORDER`
+    (asserts!
+      (< block-height (+ (get created-at preorder) NAMESPACE_PREORDER_CLAIMABILITY_TTL))
+      (err ERR_NAMESPACE_PREORDER_CLAIMABILITY_EXPIRED))
+    ;; The preorder record for this namespace will be marked as "claimed"
+    (map-set namespace-preorders
+      ((hashed-salted-namespace hashed-salted-namespace) (buyer contract-caller))
+      ((created-at (get created-at preorder)) (claimed true) (stx-burned (get stx-burned preorder))))
+    ;; The namespace will be set as "revealed" but not "launched", its price function, its renewal rules, its version,
+    ;; and its import principal will be written to the  `namespaces` table.
+    (map-set namespaces
+      ((namespace namespace))
+      ((namespace-import namespace-import)
+       (revealed-at block-height)
+       (launched-at none)
+       (namespace-version namespace-version)
+       (renewal-rule renewal-rule)
+       (price-function price-function)))
+    (ok true))))
+
+;; NAME_IMPORT
+;; Once a namespace is revealed, the user has the option to populate it with a set of names. Each imported name is given
+;; both an owner and some off-chain state. This step is optional; Namespace creators are not required to import names.
+(define-public (name-import (namespace (buff 19))
+                            (name (buff 16))
+                            (zonefile-hash (buff 20)))
+  (let (
+    (namespace-props (unwrap!
+      (map-get? namespaces ((namespace namespace)))
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (name-currently-owned (map-get? owner-name ((owner contract-caller)))))
+    (let ( 
+        (can-sender-register-name (if (is-none name-currently-owned)
```

Comment:
This behavior may be tricky to emulate now that we treat names as NFTs, though.  Open to discussion.

Reply:
> either the namespace is launched or the namespace expires.

What do you mean by namespace expires?

Reply:
Hey, sorry, didn't see this reply until now.  A namespace that is revealed but never launched will "expire" after a year, and any names imported into it will cease to exist.  The namespace can be re-preordered and re-revealed by a different key, and the names can be re-imported and re-registered.